### ISSUE TITLE
expose Sourcepos info to SyntaxHighlighterAdapters

### ISSUE THREAD

**Issue by Gankra** [kivikakk/comrak#298]
I'm working on a project that wants to provide a warning when the user uses an unknown language on a codefence, and we use [miette's fancy machinery for printing out spans in the original source](https://docs.rs/miette/latest/miette/index.html). It would be great if we could actually print out the line/col where we found the problem. It seems like comrak has that info in its AST, but the SyntaxHighlighterAdapter api doesn't expose it. comrak also buffers up intermediate results so we can't pull any "substring of the original input, do pointer comparisons" shenanigans.

There's a few different ways to introduce this kind of functionality (break the old trait methods, add new trait methods that default call the old ones, ...). Is there one you'd prefer?

**Comment by kivikakk** [comment]
👋🐰 Hiya, lovely to see you here!

Break the old trait methods! Sourcepos _should_ be supplied everywhere, it's just such a recent introduction (#232) that some of the extensibility bits added before it don't have it. Seeing as we're still pre-1.0, I'm happier making breaking changes for a more correct API now.

**Comment by Gankra** [comment]
is there an existing API in comrak you suggest using as a template for the
new one?

**Comment by kivikakk** [comment]
I think more-or-less adding the `sourcepos: Option<Sourcepos>` parameter ([e.g.](https://github.com/kivikakk/comrak/blob/2cac9d024cdfa36d7373459949be4dd8676725f4/src/adapters.rs#L66)) should be sufficient.

**Comment by Gankra** [comment]
In this case we have two things we'd want the Sourcepos of: the lang identifier and the block. Wasn't sure if you had any examples like that.

I mean I guess they're always(???) separated by a single newline so maybe you only need one..?

**Comment by Gankra** [comment]
Oh boy I am reading the code to see how complicated it is to properly find those sub-sourceposes of the NodeCodeBlock and I am quickly learning that this is a very complex parser and of course code blocks have a bunch of extra wobbly crap to adjust for (markdown parsing complex?? what a shock!!! 😸)

**Comment by Gankra** [comment]
> I'd be happy with adding either the length of the trimmed prefix (to make the rest inferable), or, actually just add an info_sourcepos to the NodeCodeBlock and record it explicitly.

Yeah I think you're probably far more qualified to make this adjustment than me. Just to be clear on what I think a user would plausibly want to know, in decreasing order:

1. the span of the body (literal?) of the code block (so a syntax highlighter can output errors as offsets within the input it processes, and that can just be directly applied to the sourcepos of the body to get an absolute span in the document)
2. the span of the info components (so you can point at it and be like "idk what this language is")
3. the span of the entire code block (what the AST stores? Although I'm not sure when I'd want to point at the whole thing)

**Comment by kivikakk** [comment]
Right, understood!

(To be clear, I don't mean to say I actually have the time/energy to do the adding of it myself! Just what I'd be fine with adding to the data model. There's been a lot of maintenance work to do on Comrak lately, and it's already unsustainable given lack of meaningful sponsorship etc.)

**Comment by Gankra** [comment]
Totally understood. I'll look over your comment a bit closer because it's definitely filling in some of the context I was missing (and I might do some driveby docs/comments to help others contribute more too).
